cmds-42/usr.bin/rcs/man/rcsclean.1
.TH RCSCLEAN L "October 30, 1989" "Purdue University"
.SH NAME
rcsclean \- clean up working files
.SH SYNOPSIS
\fBrcsclean\fR [ \fB-r\fIrev\fR ] [ \fB-q\fIrev\fR ] file...
.SH DESCRIPTION
.I Rcsclean
removes working files that were checked out and never modified.
For each file given, \fIrcsclean\fR
compares the working file and a revision in the corresponding
RCS file. If it finds no difference, it removes the working file, and,
if the revision was locked by the caller, unlocks the revision.
.PP
A file name ending in ',v' is an RCS file name, otherwise a
working file name. \fIRcsclean\fR derives the working file name from the RCS
file name and vice versa, as explained in
.IR co (L).
Pairs consisting
of both an RCS and a working file name may also be specified.
.PP
\fIRev\fR specifies with which revision the working file is compared.
If \fIrev\fR is omitted, \fIrcsclean\fR compares the working file with
the latest revision
on the default branch (normally the highest branch on the trunk).
The option \fB-q\fR suppresses diagnostics.
.PP
\fIRcsclean\fR is useful for "clean" targets in Makefiles.
Note that
.IR rcsdiff (L)
prints out the differences.
Also,
.IR ci (L)
normally asks whether to check in a file
if it was not changed.
.SH EXAMPLES
.nf
The command

	rcsclean *.c *.h

.fi
removes all working files ending in ".c" or ".h" that were not changed
since their checkout.
.SH DIAGNOSTICS
The exit status is 0 if there were no differences
during the last comparison or if the last working file did not exist,
1 if there were differences, and 2 if there were errors.
.SH IDENTIFICATION
.de VL
\\$2
..
Author: Walter F. Tichy,
Purdue University, West Lafayette, IN, 47907.
.sp 0
Revision Number:
.VL $Revision: 1.3 $
; Release Date:
.VL $Date: 89/10/30 17:36:41 $
\&.
.sp 0
Copyright \(co 1982, 1988, 1989 by Walter F. Tichy.
.SH SEE ALSO
co(L), ci(L), ident(L), rcs(L), rcsdiff(L), rcsintro(L), rcsmerge(L),
rlog(L), rcsfile(L).


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.bin/rcs/src/rcsclean.c
/*
 *                     RCS rcsclean operation
 */

#ifndef lint
static char rcsid[]=
"$Header: /usr/src/local/bin/rcs/src/RCS/rcsclean.c,v 4.5 89/10/30 12:29:00 trinkle Exp $ Purdue CS";
#endif
/*****************************************************************************
 *                       remove unneded working files
 *****************************************************************************
 */

/* Copyright (C) 1982, 1988, 1989 Walter Tichy
   Distributed under license by the Free Software Foundation, Inc.

This file is part of RCS.

RCS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

RCS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RCS; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

Report problems and direct all questions to:

    rcs-bugs@cs.purdue.edu

*/




/* $Log:	rcsclean.c,v $
 * Revision 4.5  89/10/30  12:29:00  trinkle
 * Added -q option to agree with man page, added code to actually unlock
 * the RCS file if there were no differences, picked a bit of lint.
 * 
 * Revision 4.4  89/05/01  15:12:21  narten
 * changed copyright header to reflect current distribution rules
 * 
 * Revision 4.3  88/11/08  15:59:54  narten
 * removed reference to TARGETDIR
 * 
 * Revision 4.2  87/10/18  10:30:43  narten
 * Updating version numbers. Changes relative to 1.1 are actually
 * relative to 4.1
 * 
 * Revision 1.2  87/09/24  13:59:13  narten
 * Sources now pass through lint (if you ignore printf/sprintf/fprintf 
 * warnings)
 * 
 * Revision 1.1  84/01/23  14:50:16  kcs
 * Initial revision
 * 
 * Revision 4.1  83/12/15  12:26:18  wft
 * Initial revision.
 * 
 */
#include "rcsbase.h"
#define ERRCODE 2                   /*error code for exit status            */
#ifndef lint
static char rcsbaseid[] = RCSBASE;
#endif

extern int    cleanup();            /* cleanup after signals                */
extern char * getcaller();          /* get login of caller                  */
extern struct hshentry   * findlock(); /* find and remove lock              */
extern char * mktempfile();         /*temporary file name generator         */
extern int    fterror();            /*forward for special fatal error func. */
extern struct hshentry * genrevs(); /*generate delta numbers                */
extern int    nerror;               /*counter for errors                    */
extern int    quietflag;            /*suppresses diagnostics                */
extern FILE * finptr;               /* RCS input file                       */
extern FILE * fopen();

char *RCSfilename;
char *workfilename;
char * tempfile;
char * caller;                      /* caller's login;                      */
FILE * file1, * file2;              /*file descriptors for comparison       */

main (argc, argv)
int argc; char **argv;
{
        char * cmdusage;
        char command[NCPPN+revlength+40];
	char * rev;                   /* revision number from command line  */
        char numericrev[revlength];   /* holds expanded revision number     */
	int  revnums;                 /* number of -r options               */
        struct hshentry * gendeltas[hshsize];/*stores deltas to be generated*/
        struct hshentry * target;
	register int c1;              /* reading input                      */
	int  result;                  /* result of comparison               */
	int pairresult;               /* reulst of pairfilenames            */

        catchints();
	cmdid = "rcsclean";
	cmdusage = "command format:\n    rcsclean [-rrev] [-qrev] file";
	revnums=0;
	quietflag=false;
	caller=getcaller();
        while (--argc,++argv, argc>=1 && ((*argv)[0] == '-')) {
                switch ((*argv)[1]) {
                case 'r':
		revno:	if ((*argv)[2]!='\0') {
                            if (revnums==0) {
				    rev= *argv+2; revnums=1;
                            } else {
				    fterror("too many revision numbers");
                            }
                        } /* do nothing for empty -r */
                        break;
		case 'q':
			quietflag=true;
			goto revno;

                default:
			fterror("unknown option: %s\n%s", *argv, cmdusage);
                };
        } /* end of option processing */

	if (argc<1) fterror("No input file\n%s", cmdusage);

        /* now handle all filenames */
        do {
                finptr=NULL;
		pairresult=pairfilenames(argc,argv,false,false);

		if (pairresult==0) continue; /* error */
		if (!(access(workfilename,4)==0)) {
			diagnose("Can't open %s",workfilename);
                        continue;
		} elsif (pairresult == -1) {
			warn("Can't find RCS file for %s",workfilename);
			continue;
		}
                diagnose("RCS file: %s",RCSfilename);
                if (!trysema(RCSfilename,false)) continue; /* give up */


                gettree(); /* reads in the delta tree */

                if (Head==nil) {
                        error("no revisions present");
                        continue;
                }
                if (revnums==0)
			rev=(Dbranch!=nil?Dbranch->num:Head->num); /* default rev1 */

		if (!expandsym(rev,numericrev)) continue;
                if (!(target=genrevs(numericrev,(char *)nil,(char *)nil,(char *)nil,gendeltas))) continue;

		tempfile=mktempfile("/tmp/",TMPFILE1);
		diagnose("retrieving revision %s",target->num);
                VOID sprintf(command,"%s -q -p%s %s > %s\n",
			CO ,target->num,RCSfilename,tempfile);
                if (system(command)){
                        error("co failed");
                        continue;
                }
		/* now do comparison */
		if ((file1=fopen(tempfile,"r"))==NULL) {
			error("Can't open checked out file %s",tempfile);
			continue;
		}
		if ((file2=fopen(workfilename,"r"))==NULL) {
			error("Can't open %s",workfilename);
			continue;
		}
		result=1;
		while ((c1=getc(file1))==getc(file2)) {
			if (c1==EOF) {
				/* identical files; can remove working file */
				result=0;
				diagnose("files identical; %s removed",workfilename);
				if (unlink(workfilename)!=0) {
					error("Can't unlink %s",workfilename);
				}
				if (findlock(caller, false)) {
				    VOID sprintf(command,"%s -q -u%s %s\n",
						 RCS_CMD, target->num, RCSfilename);
				    if (system(command)) {
					error("unlock failed");
					continue;
				    }
				}
				    
				break;
			}
		}
		VOID fclose(file1); VOID fclose(file2);

		if (result==1) diagnose ("files different");


        } while (cleanup(),
                 ++argv, --argc >=1);


	if (nerror>0) {
		exit(ERRCODE);
	} else {
		exit(result);
	}

}


/*VARARGS1*/
fterror(e, e1, e2)
char * e, * e1, * e2;
/* prints error message and terminates program with ERRCODE */
{       nerror++;
        VOID fprintf(stderr,"%s error: ",cmdid);
	VOID fprintf(stderr,e, e1, e2);
        VOID fprintf(stderr,"\n%s aborted\n",cmdid);
        VOID cleanup();
	exit(ERRCODE);
}

